ui/src/geoviite-design-lib/snackbar/snackbar.tsx

```diff
@@ -67,6 +73,59 @@ function getToast(headerKey: string, bodyKey?: string, button?: SnackbarButtonOp
     );
 }
 
+function getErrorToast(
+    headerKey: string,
+    errorObj?: ApiErrorResponse,
+    button?: SnackbarButtonOptions,
+) {
+    const t = i18n.t;
+
+    const header = t(headerKey);
+    const buttonText = button ? t(button.text) : undefined;
+    const copyToClipboard = () => {
+        if (errorObj) {
+            navigator.clipboard.writeText(JSON.stringify(errorObj));
+            success(t('error.copied-to-clipboard'), undefined, {
+                autoClose: COPY_CONFIRM_AUTO_CLOSE_TIMEOUT,
+            });
+        }
+    };
+
+    return (
+        <div className={styles['Toastify__toast-content']}>
+            <div className={styles['Toastify__toast-text']}>
+                <span className={styles['Toastify__toast-header-container']}>
+                    <span className={styles['Toastify__toast-header']} title={header}>
+                        {header}
+                    </span>
+                    {errorObj && (
+                        <Button
+                            title={t('error.copy-details-to-clipboard')}
+                            onClick={copyToClipboard}
+                            variant={ButtonVariant.GHOST}
+                            size={ButtonSize.SMALL}
+                            icon={Icons.Copy}
+                        />
+                    )}
+                </span>
+                {errorObj && (
+                    <div className={styles['Toastify__toast-footer']}>{`${new Date(
+                        errorObj.timestamp,
+                    ).toLocaleString(i18next.language)} | ${errorObj.correlationId}`}</div>
```

Comment:
Meillä ei kai nykytilanteessa ole palautus-jsonien rakennevalidointia clientillä, joten lienee mahdollista, että vaikka errorObj onkin määritelty, niin nämä kentät eivät olisi?

Reply:
Tuo on joo sinällään totta. Nopealla tutkinnalla tämä todennäköisesti toimisi tuollaisessa tilanteessa seuraavanlaisesti:
* Toasti tulee edelleen näkyviin (ts. sen luontikoodi ei heitä tai muutenkaan luovuta kesken matkaan)
* Ajan parsinta palauttaa Invalid Date -olion, correlationId:n parsinta `undefined`:in, jolloin käytännössä alareunaan tulee lukemaan `Invalid Date | undefined`. Ehkä tuollekin voisi jonkinlaisen oikeamman validoinnin tehdä

Reply:
Pähkäilin tätä jonkin aikaakin, mutta en oikeastaan päässyt tässä mihinkään fiksuun lopputulemaan. Nykyiselläänhän voi tosiaan käydä niin, että clientille tuleva error on jonkin muun tyyppinen kuin `ApiErrorResponse`, ja kun me vaan käytännössä oletetaan että kaikki ei-OK:t on `ApiErrorResponse`:ja, niin tyypit eivät enää päde. Tällöin päädyttäisiin tuohon `Invalid Date | undefined`-tilanteeseen.

Käytännössähän oikeastaan ainoa paikka mistä tuollaisia virheitä voisi tulla on CloudFront ja muu AWS, ja ne tuntuvat palauttavan virheviesteissä HTML-muotoisia bodyjä jotka eivät siten parsiutune JSON:iksi clientilla, joten tämä ei välttämättä ole oikeasti ongelma, mutta on se ehkä hitusen kiusallista jos niin koskaan käy. Tärkeimpänä ehkä olisi saada testauksessa varmistuttua siitä miten tämä nykyinen toteutus oikein käyttäytyy CloudFrontin palauttaessa jonkun virheen.

TL;DR: Pohdin hartaasti, en taida tehdä tässä vaiheessa tälle mitään.

Reply:
Mjeps, lienee aika harvinainen tilanne, mutta ne harvinaiset suorituksenaikaiset bugit ovatkin toisaalta niitä raskaimpia selvittää, joita täsmällisellä tyypityksellä pyrittäisiin välttämään.

Itse tosiaan käyttänyt Zodia aiemmin näihin täsmällisempiin rakennevalidaatioihin (josta taisin jo jossain aiemmassa yhteydessä mainitakkin). Tosin tällaisten käytössä on tietysti vaihtoehtoiskustannuksena ainakin devauskuormitus/-nopeus. Hyöty jäisi ainakin tässä tapauksessa melko pieneksi näin raskaasta rakennevalidaatiosta.

Jos tyypitystä haluaisi korjata (vaikka tähän snackbariin), niin `unknown` edellyttäisi varmistamaan kenttien olemassaolon ennen kuin niitä voisi käyttää. Myös typeguardia voisi käyttää tuntemattomalle tyypille jotenkin tämän suuntaisesti:
```    
const isValidErrorObj = (someErrorObj: unknown): someErrorObj is ApiErrorResponse =>
    typeof someErrorObj === 'object' &&
    someErrorObj !== null &&
    'timestamp' in someErrorObj &&
    'correlationId' in someErrorObj

...
{isValidErrorObj(errorObj) && (...)}
```
Toki sitten pitäisi vielä miettiä se mitä näytetään, jos tuo typeguardi palauttaisikin falsen 😄

Reply:
Mmh, tuollainen validaatio tänne snackbariin itseensä ei välttämättä olisi huono. Sellaisen voisin kyllä toteuttaa.